File: src/lib/telemetry/logger.ts
import pino, { Logger } from "pino";
import { trace } from "@opentelemetry/api";

export const runtime = "nodejs";

declare const EdgeRuntime: string | undefined;

type LogMethod = (obj: unknown, msg?: string, ...args: unknown[]) => void;

const sensitiveKeyPattern =
  /authorization|cookie|password|secret|token|access[_-]?token|refresh[_-]?token/i;
const credentialPattern =
  /(authorization|cookie|password|secret|token|code)=([^&\s]+)/gi;
const jwtPattern = /eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/g;

export function redactSensitiveText(value: string): string {
  return value
    .replace(credentialPattern, "$1=[REDACTED]")
    .replace(jwtPattern, "[REDACTED_JWT]");
}

function redactLogValue(value: unknown, depth = 0): unknown {
  if (typeof value === "string") return redactSensitiveText(value);
  if (value instanceof Error) {
    return {
      name: value.name,
      message: redactSensitiveText(value.message),
      stack: value.stack ? redactSensitiveText(value.stack) : undefined,
    };
  }
  if (depth > 5 || value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) {
    return value.slice(0, 100).map((item) => redactLogValue(item, depth + 1));
  }
  return Object.fromEntries(
    Object.entries(value as Record<string, unknown>).map(([key, item]) => [
      key,
      sensitiveKeyPattern.test(key)
        ? "[REDACTED]"
        : redactLogValue(item, depth + 1),
    ]),
  );
}

const isDev = process.env.NODE_ENV !== "production";
const level = isDev ? "debug" : "info";
const ENABLE_TRACING = process.env.ENABLE_OTEL_TRACING !== "false";
const IS_EDGE =
  typeof window !== "undefined" || typeof EdgeRuntime !== "undefined";

const baseConfig = {
  level,
  timestamp: pino.stdTimeFunctions.isoTime,
  redact: {
    paths: [
      "authorization",
      "cookie",
      "set-cookie",
      "password",
      "token",
      "access_token",
      "refresh_token",
      "secret",
      "*.authorization",
      "*.cookie",
      "*.password",
      "*.token",
      "*.access_token",
      "*.refresh_token",
      "*.secret",
      "req.headers.authorization",
      "req.headers.cookie",
      "res.headers.set-cookie",
    ],
    censor: "[REDACTED]",
  },
  serializers: {
    err: pino.stdSerializers.err,
  },
  ...(ENABLE_TRACING
    ? {
        mixin() {
          try {
            const span = trace.getActiveSpan();
            return span
              ? {
                  traceId: span.spanContext().traceId,
                  spanId: span.spanContext().spanId,
                }
              : {};
          } catch (err) {
            console.error("Logger base config error", err);
            return {};
          }
        },
      }
    : {}),
};

let baseLogger: Logger;
if (!IS_EDGE) {
  const { default: fs } = await import("fs");
  const { default: path } = await import("path");

  try {
    const logsDir = path.join(process.cwd(), "logs");
    if (!fs.existsSync(logsDir)) {
      fs.mkdirSync(logsDir, { recursive: true });
    }
    const logFilePath = path.join(logsDir, "app.log");
    const logStream = fs.createWriteStream(logFilePath, { flags: "a" });
    baseLogger = pino(baseConfig, logStream);
  } catch (err) {
    console.error("Failed to setup file logging, falling back to console", err);
    baseLogger = pino(baseConfig);
  }
} else {
  baseLogger = pino(baseConfig);
}

export function createLogger(name: string, projectId?: string): Logger {
  const bindings = projectId ? { name, projectId } : { name };
  const childLogger = baseLogger.child(bindings);

  if (!ENABLE_TRACING) {
    return childLogger;
  }

  const tracingLogger = pino(
    {
      ...baseConfig,
      mixin: () => {
        try {
          const span = trace.getActiveSpan();
          return span
            ? {
                traceId: span.spanContext().traceId,
                spanId: span.spanContext().spanId,
              }
            : {};
        } catch (err) {
          console.error("Logger tracing error", err);
          return {};
        }
      },
    },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    childLogger as any
  );

  const methods = ["trace", "debug", "info", "warn", "error", "fatal"] as const;
  const wrappedLogger = {} as Logger;

  methods.forEach((method) => {
    wrappedLogger[method] = (...args: Parameters<LogMethod>) => {
      args = args.map((arg) => redactLogValue(arg)) as Parameters<LogMethod>;
      try {
        const span = trace.getActiveSpan();
        if (span) {
          const { traceId, spanId } = span.spanContext();
          if (
            args.length > 0 &&
            typeof args[0] === "object" &&
            args[0] !== null
          ) {
            args[0] = { ...args[0], traceId, spanId };
          } else {
            args.unshift({ traceId, spanId });
          }
        }
      } catch (error) {
        childLogger.error({ error }, "Failed to inject tracing context");
      }
      // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
      return (childLogger[method] as Function)(...args);
    };
  });

  return Object.assign(tracingLogger, wrappedLogger, {
    child: (bindings: Record<string, unknown>) =>
      createLogger(
        `${name}:${bindings.name || "child"}`,
        bindings.project_id as string | undefined
      ),
  });
}

export const logger = createLogger("app");

File: backend/pkg/logger/logger.go
package logger

import (
	"context"
	"os"
	"regexp"
	"strings"

	"github.com/rs/zerolog"
	"go.opentelemetry.io/otel/trace"
)

var credentialPattern = regexp.MustCompile(`(?i)(authorization|cookie|password|secret|token|code)=([^&\s]+)`)
var jwtPattern = regexp.MustCompile(`eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+`)

// Logger wraps a zerolog.Logger for structured logging
type Logger struct {
	zlog *zerolog.Logger
}

type contextKey string

const (
	userIDKey    contextKey = "user_id"
	projectIDKey contextKey = "project_id"
)

// NewLogger initializes a new structured logger
func NewLogger() *Logger {
	zl := zerolog.New(os.Stdout).With().Timestamp().Logger()
	return &Logger{zlog: &zl}
}

// WithProjectID stores a project ID in a context
func WithProjectID(ctx context.Context, projectID string) context.Context {
	return context.WithValue(ctx, projectIDKey, projectID)
}

// GetProjectIDFromContext retrieves the stored project_id from the context
func GetProjectIDFromContext(ctx context.Context) (string, bool) {
	projectID, ok := ctx.Value(projectIDKey).(string)
	return projectID, ok
}

// Add common OTEL trace/span/project fields
func enrichEventWithContext(ctx context.Context, evt *zerolog.Event) *zerolog.Event {
	if projectID, ok := GetProjectIDFromContext(ctx); ok {
		evt = evt.Str("project_id", projectID)
	}

	spanCtx := trace.SpanContextFromContext(ctx)
	if spanCtx.IsValid() {
		evt = evt.
			Str("trace_id", spanCtx.TraceID().String()).
			Str("span_id", spanCtx.SpanID().String())
	}

	return evt
}

// Info logs an info-level message
func (l *Logger) Info(ctx context.Context, msg string, fields ...interface{}) {
	evt := enrichEventWithContext(ctx, l.zlog.Info())

	for i := 0; i+1 < len(fields); i += 2 {
		key, ok := fields[i].(string)
		if !ok {
			continue
		}
		evt = addSafeField(evt, key, fields[i+1])
	}
	evt.Msg(msg)
}

// Error logs an error-level message
func (l *Logger) Error(ctx context.Context, msg string, err error, fields ...interface{}) {
	evt := enrichEventWithContext(ctx, l.zlog.Error())
	if err != nil {
		evt = evt.Str("error", RedactSensitiveText(err.Error()))
	}

	for i := 0; i+1 < len(fields); i += 2 {
		key, ok := fields[i].(string)
		if !ok {
			continue
		}
		evt = addSafeField(evt, key, fields[i+1])
	}
	evt.Msg(msg)
}

// Error logs with fields
func (l *Logger) ErrorWithFields(ctx context.Context, msg string, fields map[string]interface{}) {
	evt := enrichEventWithContext(ctx, l.zlog.Error())

	for k, v := range fields {
		evt = addSafeField(evt, k, v)
	}
	evt.Msg(msg)
}

func addSafeField(evt *zerolog.Event, key string, value interface{}) *zerolog.Event {
	lowerKey := strings.ToLower(key)
	for _, sensitive := range []string{"authorization", "cookie", "password", "secret", "token", "code"} {
		if strings.Contains(lowerKey, sensitive) {
			return evt.Str(key, "[REDACTED]")
		}
	}
	if text, ok := value.(string); ok {
		return evt.Str(key, RedactSensitiveText(text))
	}
	return evt.Interface(key, value)
}

func RedactSensitiveText(value string) string {
	value = credentialPattern.ReplaceAllString(value, "$1=[REDACTED]")
	return jwtPattern.ReplaceAllString(value, "[REDACTED_JWT]")
}
